qualtran/_infra/data_types.py

```diff
@@ -0,0 +1,147 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import abc
+from typing import TypeVar, Union
+
+import attrs
+import sympy
+
+
+class QDType:
+    @property
+    @abc.abstractmethod
+    def num_qubits(self):
+        """Number of qubits required to represent a single instance of this data type."""
+
+
+QDTypeT = TypeVar("QDTypeT", bound=QDType)
+
+
+@attrs.frozen
+class QBit(QDType):
+    """A single qubit. The smallest addressable unit of quantum data."""
+
+    @property
+    def num_qubits(self):
+        return 1
+
+
+@attrs.frozen
+class QAny(QDType):
+    """Opaque bag-of-qbits type. Should be used sparingly"""
+
+    bitsize: Union[int, sympy.Expr]
+
+    def __attrs_post_init__(self):
+        if self.bitsize == 1:
+            raise ValueError("bitsize must be greater than 1. Use QBit() instead.")
+
+    @property
+    def num_qubits(self):
+        return self.bitsize
+
+
+@attrs.frozen
+class QInt(QDType):
+    """Integer of a given width bitsize.
+
+    Attributes:
+        bitsize: The number of qubits used to represent the integer.
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    def __attrs_post_init__(self):
+        if self.bitsize == 1:
+            raise ValueError("bitsize must be greater than 1. Use QBit() instead.")
+
+    @property
+    def num_qubits(self):
+        return self.bitsize
+
+
+@attrs.frozen
+class QUnsignedInt(QDType):
```

Comment:
do we need a type heirarchy? do we need an abstract QInt where is_dtype(x, QAbstractInt) returns True for QInt, QUnsignedInt, BoundedQInt?

Reply:
I think we might down the line, for example, the adder circuits should work for integer types (I think it works for fixed point too). For the moment I think it's best to see where things land given that circuit bloqs are implemented which may secretly be assuming unsigned integers etc, ditto for classical simulation.